ffi/src/engine_data.rs

```diff
@@ -93,3 +93,25 @@ fn get_raw_arrow_data_impl(data: Box<dyn EngineData>) -> DeltaResult<*mut ArrowF
     let ret_data = Box::new(ArrowFFIData { array, schema });
     Ok(Box::leak(ret_data))
 }
+
+#[cfg(feature = "default-engine")]
+#[no_mangle]
+pub unsafe extern "C" fn get_engine_data(
+    data: *mut ArrowFFIData,
+    engine: Handle<SharedExternEngine>,
+) -> ExternResult<Handle<ExclusiveEngineData>> {
+    get_engine_data_impl(data)
+        .map(|engine_data| engine_data.into())
+        .into_extern_result(&engine.as_ref())
+}
+
+#[cfg(feature = "default-engine")]
+unsafe fn get_engine_data_impl(data: *mut ArrowFFIData) -> DeltaResult<Box<dyn EngineData>> {
+    let array_local = std::ptr::read(&(*data).array);
+    let schema_local = std::ptr::read(&(*data).schema);
+
+    let array: StructArray = from_ffi(array_local, &schema_local)?.into();
+    let record_batch: delta_kernel::arrow::array::RecordBatch = array.into();
```

Comment:
```suggestion
    let data = unsafe { &*data };
    let array_data = unsafe { std::ptr::read(&data.array) };
    let array_data = unsafe { from_ffi(array_data, &data.schema) };
    let array = StructArray::from(array_data?);
    let record_batch = delta_kernel::arrow::array::RecordBatch::from(array);
```
For documentation purposes, we use tightly scoped `unsafe` blocks even within unsafe functions where the compiler doesn't require it. That said, we can probably eliminate the pointer manipulation entirely (see other comments).

Also, `from` is normally better than `into`, if type annotations are anyway needed.